src/pip/_internal/commands/list.py

```diff
@@ -160,26 +169,35 @@ def run(self, options, args):
             packages = self.get_uptodate(packages, options)
 
         self.output_package_listing(packages, options)
+        return SUCCESS
 
     def get_outdated(self, packages, options):
+        # type: (List[Distribution], Values) -> List[Distribution]
         return [
             dist for dist in self.iter_packages_latest_infos(packages, options)
             if dist.latest_version > dist.parsed_version
         ]
 
     def get_uptodate(self, packages, options):
+        # type: (List[Distribution], Values) -> List[Distribution]
         return [
             dist for dist in self.iter_packages_latest_infos(packages, options)
             if dist.latest_version == dist.parsed_version
         ]
 
     def get_not_required(self, packages, options):
-        dep_keys = set()
+        # type: (List[Distribution], Values) -> List[Distribution]
+        dep_keys = set()  # type: Set[Distribution]
         for dist in packages:
             dep_keys.update(requirement.key for requirement in dist.requires())
-        return {pkg for pkg in packages if pkg.key not in dep_keys}
+
+        # Create a set to remove duplicate packages, and cast it to a list
+        # to keep the return type consistent with get_outdated and
+        # get_uptodate
+        return list({pkg for pkg in packages if pkg.key not in dep_keys})
 
     def iter_packages_latest_infos(self, packages, options):
+        # type: (List[Distribution], Values) -> Distribution
```

Comment:
I am not sure, but should it be `Iterator[Distribution]`?

Reply:
Hi @gutsytechster 

I think the function name is misleading. Since it starts with `iter`, you might think that it takes an iterator. But it actually takes a `List[Distribution]`, being returned first by `pip._internal.utils.misc.get_installed_distributions` https://github.com/pypa/pip/pull/8016/files#diff-11a93d0cc9ae5ef29d0124a9f3775d2fR153 and then passing it out as a List to further function and then to `iter_packages_latest_infos`.

Reply:
Sorry @deveshks, I was not clear. I actually meant for return type.

Reply:
Aah yes you are correct. I just saw the `yield dist` at https://github.com/pypa/pip/pull/8016/files#diff-11a93d0cc9ae5ef29d0124a9f3775d2fR239 , because of which we need to set the return type as `Iterator[Distribution]`. Have fixed that in latest commit as well